### ISSUE TITLE
active-page-aware extent_hooks?

### ISSUE THREAD

**Issue by glenn-elliott** [jemalloc/jemalloc#2693]
I would like my extent hooks to have more visibility into the internal state of the arena. Specifically, the current number of active pages. What is the best way to go about implementing this?

I am developing an extent hook where extents are backed by large chunks of `mmap(MAP_POPULATE)`'ed memory.  My goal is to avoid page-faults after application initialization.  `jemalloc` is unaware that I have prefaulted a great deal of memory, so it greedily requests increasingly large extents that I am quickly unable to satisfy. This is usually cheap with over-commit, but that is not so in my case.

I need an informed way to tell `jemalloc` to request a smaller extent that I can fit in the memory I have already allocated. This is easy to do: I can return `NULL` from `extent_alloc_t`. However, I believe that I currently lack the ability to determine whether `jemalloc` is being greedy or if it really needs the memory. If the later, I want to try to satisfy the request by allocating another chunk of memory, since doing so lets me avoid having to impose hard memory limits on my application _a priori_.

I believe the most straightforward way to do this would be to pass my `extent_alloc_t` function the current number of active pages. I could then compare the active number of pages to the number of pages I have already `mmap()`'ed, and decide whether to `mmap()` more memory if memory is getting tight. Is there a way I can expose this to `extent_alloc_t` directly, or should I look at patching `jemalloc` to pass `shard->nactive` in `pa_alloc()` through to the lower layers of the library?

**Comment by interwq** [comment]
Try setting retain to false:
https://jemalloc.net/jemalloc.3.html#opt.retain

If the goal is to reduce page faults, change decay settings can be an asier way instead of custom hooks.

**Comment by glenn-elliott** [comment]
Thanks! I'll give `opt.retain:false` a try. The flag is multiplexing behaviors though.  From the documentation:

> It also makes jemalloc use mmap(2) or equivalent in a more greedy way, mapping larger chunks in one go.

Perhaps this behavior could be split out into a separate flag?

Actually, I think I overlooked `arena.<i>.retain_grow_limit` the last time I read the docs. This might work for me, but I would still feel more comfortable if I had a sense of how many pages were actually in use.

With regards to the decay settings, this is actually where I started, but I ran into what appears to be a memory leak.  I intend to file a bug report with a tool to reproduce, but here's a preview:

Environment:
1. Linux
2. Background garbage collector thread disabled.
3. `dirty_decay_ms` and `muzzy_decay_ms` set to `-1` (disabled)

Test:
1. Set up a dedicated arena for a Producer thread, with dirty and muzzy decay disabled. 
2. In a tight loop, the Producer thread allocates memory from its arena. _It does not touch the allocated memory._  Pointer to the data is passed to a single Consumer thread through a lockless queue.
3. In a tight loop, the Consumer thread, which does not share the arena of the Producer, removes an allocation from the queue, does not touch it, and only frees the memory.

The OOM killer kills my test program within a minute.

I believe I need to use a hook in my case, because I need to prefault the memory before my application enters its real-time event loop. Unable to accept the delay of incremental faults as the working-set size of the heap grows, even if the memory is not returned to the OS. I attempted to `malloc()` a big chunk of memory to prefault the heap, but I don't think this will work, given how jemalloc manages size classes.
